src/hotspot/share/ci/ciEnv.hpp

```diff
@@ -466,6 +467,28 @@ class ciEnv : StackObj {
   // RedefineClasses support
   void metadata_do(MetadataClosure* f) { _factory->metadata_do(f); }
 
+  // FoldStableValues support
+private:
+  class StableValue {
+    private:
+      const ciObject* _obj;
+      int _off;
+      ciConstant _value;
+
+    public:
+      StableValue() : _obj(NULL), _off(0), _value(ciConstant()) { }
+      StableValue(const ciObject* obj, int off, ciConstant value) : _obj(obj), _off(off), _value(value) { }
+
+      const ciObject* obj() const { return _obj; }
+      int off() const { return _off; }
+      ciConstant value() const { return _value; }
+  };
+
+  GrowableArray<StableValue>* _stable_values; // Cache of stable values
```

Comment:
Aren't `ciInstance` and `ciArray` provide a better place to cache field/element values (logic can be encapsulated in `ciInstance::field_value()` and `ciArray::element_value()`)?

Reply:
And it would seamlessly cover all possible cases, like static final fields (which can be concurrently updated through Reflection API or Unsafe).

Reply:
But wouldn't that require duplication of the `ciEnv::check_stable_value` logic and having a `GrowableArray` per `ciObject` instance? 

Regarding covering more cases like static final fields: I intentionally omitted them because I don't think it's possible to delay constant folding of such loads until CCP (since `null` is a valid value) and the lookup is done only once during (I)GVN. The regression test includes a corresponding case (see `testFinal`). Even with unsafe accesses, if the field holder is not known at parse time, the access is conservatively marked as mismatched and thus no constant folding is attempted later.

Reply:
> But wouldn't that require duplication of the ciEnv::check_stable_value logic and having a GrowableArray per ciObject instance?

Are you primarily concerned about footprint? The field can be lazily initialized to reduce possible effects.
`check_stable_value` still can be shared (e.g., placed on `ciObject`).

> Regarding covering more cases like static final fields: I intentionally omitted them because I don't think it's possible to delay constant folding of such loads until CCP

If you want to limit the current patch to stable case only, that's fine with me. But keep in mind that we still want to uniformly treat all cases where field/array element values are considered constant. Concurrent field modifications introduce non-determinism into compilation process which we would like to avoid. So, even if we don't cover final fields right away, let's prepare for a future enhancement which will cover that. In that respect, stable references in names are misleading.

Reply:
My concern is that putting the logic in `ciObject` adds a new field to all subclass instances while it's only rarely needed. Given that a large number of these objects are created during compilation, that adds to footprint even if the cache is initialized lazily.

I still gave this a try and another issue is that using `ciConstant` in `ciObject` creates a circular dependency (`invalid use of incomplete type`) because `ciConstant` depends on `ciObject`. We could of course duplicate the logic and put things both into `ciInstance` and `ciArray` but then a side table in `ciEnv` seems preferable to me.

Until proven otherwise, I would like to limit the current patch to stable cases because I don't think there are any issues for non-stable cases. I can rename the code from "stable" to "constant" but I'm not sure how much sense that makes given it's all limited to and guarded by `FoldStableValues`. What do you think?

Reply:
Personally, I'm not concerned about footprint increase from a field on `ciObject` class. Those represent oop constants observed by JITs and their count should be modest.

Speaking of relation between `ciConstant` and `ciObject`, `ciConstant.hpp` only refers to `ciObject*`, so the cycle can be broken by prepending `class ciObject;` in `ciConstant.hpp`. 

>  I don't think there are any issues for non-stable cases.
I'm fine with addressing only stable cases with the current patch for now, but I don't agree there's nothing to improve for non-stable cases.

Concurrent field modifications may seem benign, but our previous experience shows that it's safer to work with a consistent view through CI (when it is feasible/possible). 
(There's a relevant RFE filed: https://bugs.openjdk.org/browse/JDK-8294616) And it helps improve compilation replay along the way. And it's not specific to C2.

So, IMO in the longer term CI is the right place to fix it (irrespective whether stable or all constant cases are handled). It's a bit strange to see the new checks added in `opto/type.cpp` (irrespective of where the cache is located).

Reply:
> Speaking of relation between ciConstant and ciObject, ciConstant.hpp only refers to ciObject*, so the cycle can be broken by prepending class ciObject; in ciConstant.hpp.

That won't help. All `ci*` classes, including `ciObject`, are already forward declared in `ciClassList.hpp` which is included in `ciConstant.hpp`. The following usages in `ciConstant` require a complete `ciObject` type:
https://github.com/openjdk/jdk/blob/c5c8c0644d9442846de15422285fffeb91c3e0a1/src/hotspot/share/ci/ciConstant.hpp#L115
https://github.com/openjdk/jdk/blob/c5c8c0644d9442846de15422285fffeb91c3e0a1/src/hotspot/share/ci/ciConstant.hpp#L134

Another issue is that `ciConstant.hpp` includes `ciNullObject.hpp` which inherits from `ciObject`. I worked around both issues by moving code and includes around.

I moved the code to `ciObject`, as you proposed, and extended it to include final fields as well. Let me know what you think.

Testing now revealed a new "Not monotonic" assert that seems to be triggered by caching non-stable final fields. I'll investigate but I'm not able to reproduce it yet.